grr/server/grr_response_server/gui/api_plugins/metadata.py

```diff
@@ -51,14 +60,546 @@ class ApiGetOpenApiDescriptionHandler(api_call_handler_base.ApiCallHandler):
+    _ExtractSchema which takes care of error-verifications and caching.
+    """
+    t_name = self._GetTypeName(t)
+
+    schema_obj = dict()
+    schema_obj["type"] = "object"
+    schema_obj["properties"] = dict()
+    visiting.add(t_name)
+
+    for field_descriptor in t.fields:
+      field_name = field_descriptor.name
+      message_descriptor = field_descriptor.message_type # None if not Message.
+      enum_descriptor = field_descriptor.enum_type # None if not Enum.
+      descriptor = message_descriptor or enum_descriptor
+
+      if descriptor:
+        self._ExtractSchema(descriptor, visiting)
+
+      schema_or_ref_obj = \
+        self._GetSchemaOrReferenceObject(
+          self._GetTypeName(field_descriptor),
+          field_descriptor.label == protobuf2.LABEL_REPEATED)
+
+      schema_obj["properties"][field_name] = schema_or_ref_obj
+
+    visiting.remove(t_name)
+    self.schema_objs[t_name] = schema_obj
+
+    return schema_obj
+
+  def _ExtractSchema(self, t, visiting: Set[str]):
+    if t is None:
+      raise ValueError(f"Trying to extract schema of None.")
+
+    t_name = self._GetTypeName(t)
+    # "Primitive" types should be already present in self.schema_objs.
+    if t_name in self.schema_objs:
+      return self.schema_objs[t_name]
+
+    if t_name in visiting:
+      # Dependency cycle.
+      return
+
+    if isinstance(t, Descriptor):
+      return self._ExtractMessageSchema(t, visiting)
+
+    if isinstance(t, EnumDescriptor):
+      return self._ExtractEnumSchema(t, visiting)
+
+    raise TypeError(f"Don't know how to handle type \"{t_name}\" "
+                    f"which is not a protobuf message Descriptor, "
+                    f"nor an EnumDescriptor, nor a primitive type.")
+
+  def _ExtractSchemas(self):
+    """Extracts OpenAPI schemas for all the used protobuf types."""
+
+    self.schema_objs = dict()  # Holds OpenAPI representations of types.
+    self._AddPrimitiveTypesSchemas()
+
+    visiting = set()  # Holds state of types extraction (white/gray nodes).
+    router_methods = self.router.__class__.GetAnnotatedMethods()
+    for method_metadata in router_methods.values():
+      args_type = method_metadata.args_type
+      if args_type:
+        if inspect.isclass(args_type) \
+            and issubclass(args_type, rdf_structs.RDFProtoStruct):
+          self._ExtractSchema(args_type.protobuf.DESCRIPTOR, visiting)
+        else:
+          self._ExtractSchema(args_type, visiting)
+
+      result_type = method_metadata.result_type
+      if result_type:
+        if inspect.isclass(result_type) \
+            and issubclass(result_type, rdf_structs.RDFProtoStruct):
+          self._ExtractSchema(result_type.protobuf.DESCRIPTOR, visiting)
+        else:
+          self._ExtractSchema(result_type, visiting)
+
+  def _SetComponents(self):
+    if self.schema_objs is None:
+      raise ValueError("Called _SetComponents before extracting schemas.")
```

Comment:
That is not really a `ValueError`. More like an `AssertionError` and it leaks implementation details `_SetComponents` name.

Reply:
Changed the type of the exception and the message for this case and all other similar cases in the code. Please let me know if they are not correct. Thank you!